### DISCUSSION TITLE
Support for Magento2

### DISCUSSION THREAD

**Discussion by nerdeveloper** [paketo-buildpacks/feedback#204] [Web Servers Team]
I have been trying to set up Magento, but I need to figure out what I am doing wrong; it keeps going to 502 bad gateway.

I am trying to replicate this: https://devdocs.magento.com/guides/v2.3/install-gde/prereq/nginx.html

When I try to set up a `.nginx.conf.d/magento-server.conf`

I added

```bash
upstream fastcgi_backend {
  server  unix:/run/php/php-fpm.sock;
}

server {

  listen {{port}};
  server_name localhost;
  set $MAGE_ROOT /workspace;
  include /workspace/nginx.conf.sample;
}
```
I keep getting 

```bash
023/07/31 12:10:02 [emerg] 26#0: "server" directive is not allowed here in /workspace/.nginx.conf.d/magento-server.conf:1
process nginx exited, status: exit status 1
error running procs: exit status 1
```

 my run command is 
 `pack build magento --builder paketobuildpacks/builder:0.2.443-full --path .`
 
 My repo is https://github.com/nerdeveloper/magento-sample

**Comment by robdimsdale** [comment]
cc @paketo-buildpacks/web-servers

**Reply by nerdeveloper** [reply]
thank you @robdimsdale i have changed the category

**Comment by arjun024** [comment]
I have moved the discussion the PHP category as the question deals with the nginx conf generated by the [php-nginx buildpack](https://github.com/paketo-buildpacks/php-nginx).

I tried to quickly run your app but I can't get to your error as the build errored out with `The 'https://repo.magento.com/archives/magento/composer-root-update-plugin/ magento-composer-root-update-plugin-2.0.3.0.zip' URL required authentication (HTTP 401).`

Regardless, I don't think you can add server blocks to `.nginx.conf.d/*-server.conf` -> See where they are injected in the buildpack - [link](https://github.com/paketo-buildpacks/php-nginx/blob/v0.3.1/assets/nginx.conf#L208). User-provided `http` confs mean you can inject them at the http block, and user-provided `server` confs mean you can inject them inside the server block (which is inside the http block).

In case of serving php apps with nginx using paketo buildpacks, you don't really need to define your `server` as it's already set up for you. 

Your `.nginx.conf.d/magento-server.conf` should look something like:
```
    set $MAGE_ROOT /workspace;
    include /workspace/nginx.conf.sample;
```
Let us know if this doesn't help you get moving

**Reply by nerdeveloper** [reply]
I am going to try this and  for the keys; you need: https://devdocs.magento.com/guides/v2.3/install-gde/prereq/connect-auth.html

**Comment by arjun024** [comment]
The docs are here (https://paketo.io/docs/reference/php-reference/#php-nginx-buildpackbpphp-nginx-configuration) but it doesn't clearly say where the includes files will go in the main config. I have tried to clarify that in the docs https://github.com/paketo-buildpacks/paketo-website/pull/713

**Comment by nerdeveloper** [comment]
@arjun024 How about the  upstream, how do i add that 
```
upstream fastcgi_backend {
  server  unix:/run/php/php-fpm.sock;
}
```

**Comment by arjun024** [comment]
If you see the [link](https://github.com/paketo-buildpacks/php-nginx/blob/v0.3.1/assets/nginx.conf) I sent you earlier of the nginx.conf generated, it already sets the address of a FastCGI server to PHP fpm socket for you. Is there something else you're trying to achieve?

**Comment by nerdeveloper** [comment]
when i try to run it, i get this error
```bash
2023/08/01 10:04:59 [emerg] 26#0: host not found in upstream "php-fpm" in /workspace/nginx.conf.sample:50
process nginx exited, status: exit status 1
error running procs: exit status 1
```

this is line 50 code

```bash
location ~* ^/setup($|/) {
    root $MAGE_ROOT;
    location ~ ^/setup/index.php {
        fastcgi_pass   php-fpm:9000; # this is line 50
        include        fastcgi_params;

        fastcgi_param  PHP_FLAG  "session.auto_start=off \n suhosin.session.cryptua=off";
        fastcgi_param  PHP_VALUE "memory_limit=756M \n max_execution_time=600";
        fastcgi_read_timeout 600s;
        fastcgi_connect_timeout 600s;

        fastcgi_index  index.php;
        fastcgi_param  SCRIPT_FILENAME  $document_root$fastcgi_script_name;

    }
   ```

**Reply by arjun024** [reply]
Try `fastcgi_pass    php_fpm;`
Please refer to the template I linked above

**Comment by nerdeveloper** [comment]
I did that, I have a new error

```bash
2023/08/01 21:05:56 [emerg] 26#0: no port in upstream "php-fpm" in /workspace/nginx.conf.sample:77
process nginx exited, status: exit status 1
error running procs: exit status 1
```

```bash
# PHP entry point for update application
location ~* ^/update($|/) {
    root $MAGE_ROOT;

    location ~ ^/update/index.php {
        fastcgi_split_path_info ^(/update/index.php)(/.+)$;
        fastcgi_pass   php-fpm; # this is line 77
        fastcgi_index  index.php;
        fastcgi_param  SCRIPT_FILENAME  $document_root$fastcgi_script_name;
        fastcgi_param  PATH_INFO        $fastcgi_path_info;
        include        fastcgi_params;
    }
    ```

**Comment by arjun024** [comment]
@nerdeveloper underscore, not dash (`php_fpm`). See the [template](https://github.com/paketo-buildpacks/php-nginx/blob/v0.3.2/assets/nginx.conf#L132-L134) again

Imagine you provide the following addition conf in [the test app you see here](https://github.com/paketo-buildpacks/php-nginx/tree/v0.3.2/integration/testdata/default_app)
```
 ‣ cat .nginx.conf.d/user-server.conf
        location ~* \.mytest$ {
            fastcgi_param  REQUEST_METHOD  $request_method;
            fastcgi_param  SCRIPT_FILENAME $document_root/index.php;
            fastcgi_pass   php_fpm;
        }
```

then if you hit `0.0.0.0:8080/hey.mytest`, the nginx will proxy your request to the PHP PFM.
I hope that clears things up

**Comment by nerdeveloper** [comment]
OMG! Thank you so much; it works now. @arjun024 I will close this issue
